library/pkg/executor/executor.go

```diff
@@ -0,0 +1,359 @@
+// In case the process setup was successful a Command will be returned
+// The Command will either have "running" or "failed" state based on whether the process was succesfully started
+// Further interaction with the process should happen through the returned Command instance
+func (e Executor) StartJob(processId string, command string, args []string) (Command, error) {
+	err := e.createCgroup(processId)
+
+	if err != nil {
+		cleanCgroups(processId)
+		return nil, err
+	}
+
+	cmd := exec.Command(command, args...)
+
+	outputBuffer := &internal.MultiplexingBuffer{}
+
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = outputBuffer
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags: syscall.CLONE_NEWUTS | syscall.CLONE_NEWPID | syscall.CLONE_NEWUSER | syscall.CLONE_NEWNET | syscall.CLONE_NEWNS,
+		UidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getuid(),
+				Size:        1,
+			},
+		},
+		GidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getgid(),
+				Size:        1,
+			},
+		},
+	}
+
+	err = cmd.Start()
+
+	if err != nil {
+		return &internal.FailedCommand{}, nil
+	}
+
+	pid := cmd.Process.Pid
+
+	err = e.addPidToCgroups(processId, pid)
+	if err != nil {
+		killErr := cmd.Process.Kill()
+
+		if killErr != nil {
+			return nil, fmt.Errorf("the process could not be added to cgroup and killed in cleanup %w", killErr)
+		}
+
+		return nil, fmt.Errorf("failed to add process to cgroup %w", killErr)
+	}
+
+	c := &internal.StartedCommand{
+		Cmd:          cmd,
+		CmdStatus:    internal.RUNNING,
+		OutputBuffer: outputBuffer,
+		AfterDone: func() {
+			cleanCgroups(processId)
+		},
+	}
+	return c, nil
+}
+
+func cleanCgroups(processId string) {
+	processCgroupFolder := filepath.Join(cgroupFolder, processId)
+
+	_ = syscall.Rmdir(processCgroupFolder)
+}
+
+func (e Executor) createCgroup(processId string) error {
+	processCgroupFolder := filepath.Join(cgroupFolder, processId)
+	err := os.Mkdir(processCgroupFolder, 0600)
+
+	if err != nil {
+		return fmt.Errorf("failed to create cgroup %w", err)
+	}
+
+	err = os.WriteFile(filepath.Join(processCgroupFolder, "cpu.max"), []byte(e.cgroupsInfo["cpu"]), os.ModeAppend)
+	if err != nil {
+		return fmt.Errorf("failed to add resouce control %w", err)
+	}
+
+	err = os.WriteFile(filepath.Join(processCgroupFolder, "memory.max"), []byte(e.cgroupsInfo["memory"]), os.ModeAppend)
+	if err != nil {
+		return fmt.Errorf("failed to add resouce control %w", err)
+	}
+
+	err = os.WriteFile(filepath.Join(processCgroupFolder, "memory.high"), []byte(e.cgroupsInfo["memory_high"]), os.ModeAppend)
+	if err != nil {
+		return fmt.Errorf("failed to add resouce control %w", err)
+	}
+
+	diskLines, err := readDisks()
+	if err != nil {
+		return err
+	}
+
+	for _, diskLine := range diskLines {
+		parsedDiskLine := strings.Split(diskLine, "\t")
```

Comment:
Nit: Feels like this parsing logic can be tucked inside readDisks so it returns info that's already ready to use by this function.

Reply:
fair